### DISCUSSION TITLE
Dynamic logger

### DISCUSSION THREAD

**Discussion by ulkgv** [odygrd/quill#268] [Q&A]
Is there a way to create a logger/handler for use in a class instance, such that when the instance is created, a new FileHandler is used, and when destroyed, it also removes the logger/handler, saving the log? Each instance would log to a different file.

A workaround for me, in case the above is not possible, would be to manually rotate a RotatingFileHandler, so that each "session" ends up in a separated file.

Is any of the above possible?

**Comment by odygrd** [comment]
hey,

sure, it is safe to do something like this :

```c++
#include "quill/Quill.h"

class MyClass
{
public:
  MyClass(std::string const& name)
  {
    quill::Handler* file_handler = quill::file_handler(name + ".log", "w");
    _logger = quill::create_logger(name, file_handler);
  }

  void foo()
  {
    LOG_INFO(_logger, "log from foo");
  }

private:
  quill::Logger* _logger;
};

int main()
{
  // Start the logging backend thread
  quill::start();

  std::vector<MyClass> v;
  for (size_t i = 0; i < 4; ++i)
  {
    v.emplace_back("my_class" + std::to_string(i));
    v.back().foo();
  }
}
```

Currently removing the loggers/handlers it is not implemented due to complexity, those objects do not take much memory so you should be fine unless you want to create millions of class instances and files. I can probably give it a try and add a remove logger function over the next few weeks. It is a bit complex because the backend logging thread should first process everything for that logger before removing it for example

**Reply by odygrd** [reply]
I have done the logger/handlers removal in a local branch. I just need to clean it up a bit and also fix the tests as that change changes the API a bit in Quill.h
It should be ready in a few days or like 1-2 weeks

**Reply by ulkgv** [reply]
Will the handler be linked to the logger via smart pointer memory management, or the handler will also require some management so the files are flushed and saved?

**Reply by odygrd** [reply]
You don't require any management only to call `remove_logger`

Handler will be linked to logger via a shared pointer.
When all loggers that own the same handler are destroyed the handler then gets destroyed closing the file.

When you call `remove_logger` it is not removed but marked as invalidated and you should not log anymore using that `logger*`
When the backend logging thread has processed all the messages it goes garbage collection removing all invalidated loggers.

**Reply by odygrd** [reply]
Hey @ulkgv , this is now done on https://github.com/odygrd/quill/pull/277 I have also added an example https://github.com/odygrd/quill/blob/remove_loggers/examples/example_file_callbacks.cpp You don't need `FileEventNotifier ` it is there for demonstration. If there are test cases failures I should be able to fix over the next few days. Let me know if you have any issues if you decide to use it
